### PR TITLE
Clowns can now make balloon... toys. And also mallets and hats.

### PR DESCRIPTION
**Pull request by YesterdaysPromise**
## About The Pull Request

Clowns will now start with a box of 24 random long balloons and a skillchip in their noggin allowing them to create balloon animals by combining two of them of different colour together. Owners of the skillchip also gain access to crafting recepies of balloon mallets, vests, helmets and tophats, all created from long balloons. A crate of long balloons, with a box of balloons inside, can be bought at cargo, in case the clown runs out. I might edit this once I wake up, its 3 in the morning right now.
Oh also, resprited how balloons look in inventory.

image

## Why It's Good For The Game

Balloon animals funny.
Silly features are my favourite kind of features, and this one's open-ended too.
Someone on the coder chat recommended someone would do it that one time, here it goes.

## Changelog

:cl:
add: Added long balloon box to the clown's starting inventory, and a skill-chip of long lost honk-motherian knowledge to their brain.
add: Added long balloons. Consequently, added balloon animals to make from such balloons. Also, balloon top hat, vest, helmet, and a mallet. Don't ask about the mallet.
add: A long balloons box harvested fresh from the farms on the clown planet will be able to be shipped in a crate to the cargo department near you!
add: As per requests; water balloons can now be printed at service lathe, and entertainment modsuit can now blow long balloons!
image: Balloons will now have an unique sprite when in the inventory, compared when to on the ground.
/:cl:

### CODE REVIEW

**Comment by Astrogem2** [comment]
and then when you the balloon to someone you comically pop it in their face, leaving a small latex flap on it.

File: code/game/objects/items/weaponry.dm [hunk 1]
```diff
@@ -40,6 +40,48 @@ for further reading, please see: https://github.com/tgstation/tgstation/pull/301
 	if(user.combat_mode)
 		return ..(M, user)
 
+
+/obj/item/balloon_mallet
+	name = "balloon mallet"
+	desc = "It's a mallet, a weapon known for beign heavy, but made from notoriously light balloons. Air inside removes any force from the swings. It'd be quite embarassing to get hit by this."
```
Comment (necromanceranne) [inline code/game/objects/items/weaponry.dm:46]:
```suggestion
	desc = "It's a mallet, a weapon known for being heavy, but made from notoriously light balloons. Air inside removes any force from the swings. It'd be quite embarrassing to get hit by this."
```

File: code/game/objects/items/weaponry.dm [hunk 2]
```diff
@@ -40,6 +40,48 @@ for further reading, please see: https://github.com/tgstation/tgstation/pull/301
...
+	attack_verb_continuous = list("mallets", "smoother")
+	attack_verb_simple = list("mallet", "smoother")
+	max_integrity = 20
+	armor_type = /datum/armor/item_banhammer
+	resistance_flags = FIRE_PROOF
+
+/obj/item/balloon_mallet/examine(mob/user)
+	. = ..()
+	if(HAS_TRAIT(user,TRAIT_BALLOON_SUTRA))
+		. = "A sacred weapon of the higher castes from the clown planet, used to strike fear into the hearths of their foes. Wield it with care."
```
Comment (necromanceranne) [inline code/game/objects/items/weaponry.dm:67]:
```suggestion
		. = "A sacred weapon of the higher castes from the clown planet, used to strike fear into the hearts of their foes. Wield it with care."
```

**Comment by Dawnseer** [comment]
The text for the balloon lizard goes 'an lizard'. Might wanna switch that to 'a lizard'

**Review by 00-Steven — CHANGES_REQUESTED** [review]
Neat feature!
Did a quick general spelling/grammar/such passover.

Haven't looked much at the _implementation_ yet, but `/obj/item/toy/balloon/long/attackby(...)` probably definitely needs some good looking at.

File: code/game/objects/items/toys.dm [hunk 3]
```diff
@@ -127,6 +127,73 @@
...
+	if(isturf(loc))
+		icon_state = "[world_state]"
+	else
+		icon_state = "[storage_state]"
+	update_appearance()
+
+/obj/item/toy/balloon/long/attackby(obj/item/attacking_item, mob/living/user, params)
+	if(istype(attacking_item, /obj/item/toy/balloon/long) && HAS_TRAIT(user, TRAIT_BALLOON_SUTRA))
+		var/obj/item/toy/balloon/long/hit_by = attacking_item
+		if(hit_by.current_color != current_color)
```
Comment (san7890) [inline code/game/objects/items/toys.dm:152]: holy fucking shit this if chain is absolutely horrific. here's how we can make it better. first, early returns to make the indentation less bad

```dm
/obj/item/toy/balloon/long/balloon/attackby(...)
 if(!istype(attacking_item, /obj/item/toy/balloon/long) || !HAS_TRAIT(user, TRAIT_BALLOON_SUTRA)
  return ..() // saves indentation while doing the exact same thing.
 
 var/obj/item/toy/balloon/long/hit_by = attacking_item
 if(hit_by.current_color == current_color)
  to_chat(user, span_warning("You must use balloons of different colours to do that!"))
  return ..()

 // then you would have the rest of the code (visible_message, etc.) here on this indent line and qdel both items and return TRUE at the very end.
```

we'll need to do more work, but start here. you would de-indent all the coming lines to just make it slightly less insane.

File: code/modules/library/skill_learning/job_skillchips/clown.dm [hunk 4]
```diff
@@ -0,0 +1,10 @@
+/obj/item/skillchip/job/clown
+	name = "B@L00NY skillchip"
+	desc = "This biochip contain several terabytes of uncannily religious, Honkmother praising guides on how to reshape balloons into silly animals."
+	auto_traits = list(TRAIT_BALLOON_SUTRA)
+	skill_name = "Balloon Sutra"
+	skill_description = "Learn the the ancient Honkmotherian arts of balloon-sutra."
+	skill_icon = "face-grin-tears"
+	activate_message = "<span class='notice'>Blessed wisdom of Honkmother enwraps you, and with it, governship upon form of balloonkind.</span>"
+	deactivate_message = "<span class='notice'>'Remember, then, that true clownery requires freedom and willingness to bend, like ones of a floating balloon.'... Whatever that meant?</span>"
```
Comment (san7890) [inline code/modules/library/skill_learning/job_skillchips/clown.dm:9]: use the `span_notice()` macro.

```suggestion
	activate_message = span_notice("Blessed wisdom of Honkmother enwraps you, and with it, governship upon form of balloonkind.")
	deactivate_message = span_notice("'Remember, then, that true clownery requires freedom and willingness to bend, like ones of a floating balloon.'... Whatever that meant?")
```

File: code/game/objects/items/toys.dm [hunk 5]
```diff
@@ -200,6 +277,161 @@
 
 #undef BALLOON_COLORS
 
+/*
+* Balloon animals
+*/
+
+/obj/item/toy/balloon_guy
```
Comment (san7890) [inline code/game/objects/items/toys.dm:284]: I don't know how we can make the earlier code less ugly (yet) but a good first step would be to move all of the toys you've added to a new type, maybe `/obj/item/toy/character_balloon` and then you would make all of these toys subtypes of that type (e.g. `/obj/item/toy/character_balloon/balloon_guy`, `character_balloon/balloon_nukie`) so we can get a consistent framework going with these interactions.

**Review by san7890 — CHANGES_REQUESTED** [review]
this PR will need some work on the code side because having a 30 line `if()` conditional block is not the answer. here's some advice on how to clean it up and hopefully provide some hints on how you can create a better option (any wandering coders/maintainers welcome to provide advice).

**Comment by YesterdaysPromise** [comment]
I haven't really learned much from this, crap's just been breaking for no discernable reason, I keep whineing about it to others, and get handed solutions, which I accept but do not quite swallow. I feel really shitty.

**Comment by ophaq** [comment]
Holy shit finally someone did this. I've been wanting more stuff like this for a while now. We seriously need more clown craftable things that make them more clown-like. It would be cool if we could order more/print more in service lathe and fill the balloons with water too which makes people slip on the puddle when it's popped + makes them wet.

File: code/game/objects/items/toys.dm [hunk 6]
```diff
@@ -147,53 +147,54 @@
...
-			if((hit_by.current_color == "yellow" && current_color == "purple") || (hit_by.current_color == "purple" && current_color == "yellow"))
-				user.put_in_hands(new /obj/item/toy/balloon_animal/ethereal)
-			if((hit_by.current_color == "orange" && current_color == "purple") || (hit_by.current_color == "purple" && current_color == "orange"))
-				user.put_in_hands(new /obj/item/toy/balloon_animal/plasmaman)
-
-			qdel(hit_by)
-			qdel(src)
-			return TRUE
+	if(!istype(attacking_item, /obj/item/toy/balloon/long) || !HAS_TRAIT(user, TRAIT_BALLOON_SUTRA)
+ 		return ..() // saves indentation while doing the exact same thing.
```
Comment (san7890) [inline code/game/objects/items/toys.dm:151]: delete this comment it was just there to teach you

File: code/game/objects/items/toys.dm [hunk 7]
```diff
@@ -147,53 +147,54 @@
...
-				user.put_in_hands(new /obj/item/toy/balloon_animal/moth)
-			if((hit_by.current_color == "yellow" && current_color == "purple") || (hit_by.current_color == "purple" && current_color == "yellow"))
-				user.put_in_hands(new /obj/item/toy/balloon_animal/ethereal)
-			if((hit_by.current_color == "orange" && current_color == "purple") || (hit_by.current_color == "purple" && current_color == "orange"))
-				user.put_in_hands(new /obj/item/toy/balloon_animal/plasmaman)
-
-			qdel(hit_by)
-			qdel(src)
-			return TRUE
+	if(!istype(attacking_item, /obj/item/toy/balloon/long) || !HAS_TRAIT(user, TRAIT_BALLOON_SUTRA)
```
Comment (san7890) [inline code/game/objects/items/toys.dm:150]: parenthesis are unbalanced here

File: code/game/objects/items/toys.dm [hunk 8]
```diff
@@ -200,6 +278,96 @@
...
+	icon_state = "balloon_guy"
+
+/obj/item/toy/balloon_animal/nukie
+	name = "balloon nukie"
+	desc = "A balloon effigy of syndicate's nuclear operative. Either made to appease them and pray for survival, or to poke fun at them."
+	icon_state = "balloon_nukie"
+
+/obj/item/toy/balloon_animal/clown
+	name = "balloon clown"
+	desc = "A balloon clown, smiling from ear to ear and beyond!"'
```
Comment (san7890) [inline code/game/objects/items/toys.dm:307]:
```suggestion
	desc = "A balloon clown, smiling from ear to ear and beyond!"
```

**Review by san7890 — CHANGES_REQUESTED** [review]
this is why CI is failing

**Comment by Admiral-Gigaton** [comment]
Can you make it so when the mallet is hit on the ehatchimacallit plate that you hit a gavel on, it plays a sound effect? Your choice

File: code/game/objects/items/toys.dm [hunk 9]
```diff
@@ -127,6 +127,74 @@
...
+/obj/item/toy/balloon/long/attackby(obj/item/attacking_item, mob/living/user, params)
+	if(!istype(attacking_item, /obj/item/toy/balloon/long) || !HAS_TRAIT(user, TRAIT_BALLOON_SUTRA))
+		return ..()
+
+	var/obj/item/toy/balloon/long/hit_by = attacking_item
+	if(hit_by.current_color == current_color)
+		to_chat(user, span_warning("You must use balloons of different colours to do that!"))
+		return ..()
+
+	if(hit_by.current_color != current_color)
```
Comment (san7890) [inline code/game/objects/items/toys.dm:158]: this check is redundant and should be removed since we already properly check for this on line 154.

**Review by san7890 — CHANGES_REQUESTED** [review]

**Comment by NotSomeoneyUKnow** [comment]
balloon esword/sword and this will be epic (if you make them deal stun damage this will be amazeing)

**Comment by GuillaumePrata** [comment]
With all my hours on Don't Starve Together it makes me quite sad that the mines can't join in the fun, tho it is nice to have more distinctive features between both jobs.

Would be nice to add a balloon skillchip to the crate tho, for all Wes wannabes or any other crewmember that might want to have some fun with the balloons, we also just need more reasons to use the skillchip machine...

**Comment by Dawnseer** [comment]
Makes me wish that the balloon blowing modsuit module wasn't a maintenance module, but instead part of the entertaintment modsuit pack, which currently has a grand whopping... A waddle module, a bikehorn module, and a microwave module. And of course the entertainment plating.

File: code/game/objects/items/toys.dm [hunk 10]
```diff
@@ -127,6 +127,64 @@
...
+			list("orange", "purple") = /obj/item/toy/balloon_animal/plasmaman,
+		)
+		for(var/list/pair_of_colors in balloon_combos)
+			if((hit_by.current_color == pair_of_colors[1] && current_color == pair_of_colors[2]) || (current_color == pair_of_colors[1] && hit_by.current_color == pair_of_colors[2]))
+				var/path_to_spawn = balloon_combos[pair_of_colors] 
+				user.put_in_hands(new path_to_spawn)
+				break 
+		qdel(hit_by)
+		qdel(src)
+		return TRUE
```
Comment (san7890) [inline code/game/objects/items/toys.dm:187]:
```suggestion
	visible_message(
		span_notice("[user.name] starts contorting up a balloon animal!"),
		blind_message = span_hear("You hear balloons being contorted."),
		vision_distance = 3,
		ignored_mobs = user,
	)
	var/list/balloon_combos = list(
		list("red", "blue") = /obj/item/toy/balloon_animal/guy,
		list("red", "green") = /obj/item/toy/balloon_animal/nukie,
		list("red", "yellow") = /obj/item/toy/balloon_animal/clown,
		list("red", "orange") = /obj/item/toy/balloon_animal/cat,
		list("red", "purple") = /obj/item/toy/balloon_animal/fly,
		list("blue", "green") = /obj/item/toy/balloon_animal/podguy,
		list("blue", "yellow") = /obj/item/toy/balloon_animal/ai,
		list("blue", "orange") = /obj/item/toy/balloon_animal/dog,
		list("blue", "purple") = /obj/item/toy/balloon_animal/xeno,
		list("green", "yellow") = /obj/item/toy/balloon_animal/banana,
		list("green", "orange") = /obj/item/toy/balloon_animal/lizard,
		list("green", "purple") = /obj/item/toy/balloon_animal/slime,
		list("yellow", "orange") = /obj/item/toy/balloon_animal/moth,
		list("yellow", "purple") = /obj/item/toy/balloon_animal/ethereal,
		list("orange", "purple") = /obj/item/toy/balloon_animal/plasmaman,
	)
	for(var/list/pair_of_colors in balloon_combos)
		if((hit_by.current_color == pair_of_colors[1] && current_color == pair_of_colors[2]) || (current_color == pair_of_colors[1] && hit_by.current_color == pair_of_colors[2]))
			var/path_to_spawn = balloon_combos[pair_of_colors] 
			user.put_in_hands(new path_to_spawn)
			break 
	qdel(hit_by)
	qdel(src)
	return TRUE
```

* no `else`, you only need to deindent. also proper multilining of list calls
* also trailing commas rule

File: code/modules/cargo/packs/costumes_toys.dm [hunk 11]
```diff
@@ -268,3 +268,15 @@
...
 	crate_type = /obj/structure/closet/crate/wooden
 	discountable = SUPPLY_PACK_STD_DISCOUNTABLE
+
+/datum/supply_pack/costumes_toys/balloons
+	name = "Long Balloons Kit"
+	desc = "This crate contains a box of long balloons, plus a skillchip for non-clowns to join the fun! Extra layer of safety so clowns at CentCom won't get to them."
+	cost = CARGO_CRATE_VALUE * 4
+	contains = list(
+		/obj/item/storage/box/balloons
+		/obj/item/skillchip/job/clown
```
Comment (san7890) [inline code/modules/cargo/packs/costumes_toys.dm:278]:
```suggestion
		/obj/item/storage/box/balloons,
		/obj/item/skillchip/job/clown,
```

this is why your compile is failing

**Review by san7890 — CHANGES_REQUESTED** [review]

**Comment by YesterdaysPromise** [comment]
I hope everything is alright.

File: code/game/objects/items/weaponry.dm [hunk 12]
```diff
@@ -40,6 +40,48 @@ for further reading, please see: https://github.com/tgstation/tgstation/pull/301
...
+	max_integrity = 20
+	armor_type = /datum/armor/item_banhammer
+	resistance_flags = FIRE_PROOF
+
+/obj/item/balloon_mallet/examine(mob/user)
+	. = ..()
+	if(HAS_TRAIT(user,TRAIT_BALLOON_SUTRA))
+		. = "A sacred weapon of the higher castes from the clown planet, used to strike fear into the hearts of their foes. Wield it with care."
+
+/obj/item/balloon_mallet/attack(mob/living/M, mob/living/user)
```
Comment (jlsnow301) [inline code/game/objects/items/weaponry.dm:69]: please change `M` to something like `target`

File: code/game/objects/items/toys.dm [hunk 13]
```diff
@@ -127,6 +127,64 @@
...
+	if(hit_by.current_color == current_color)
+		to_chat(user, span_warning("You must use balloons of different colours to do that!"))
+		return ..()
+	visible_message(
+		span_notice("[user.name] starts contorting up a balloon animal!"),
+		blind_message = span_hear("You hear balloons being contorted."),
+		vision_distance = 3,
+		ignored_mobs = user,
+	)
+	var/list/balloon_combos = list(
```
Comment (Jacquerel) [inline code/game/objects/items/toys.dm:163]: Move this list into the object properties and add `static`, we don't want to recreate it every time this proc gets run.

Comment (YesterdaysPromise) [reply code/game/objects/items/toys.dm:163]: what do you mean?

File: code/game/objects/items/toys.dm [hunk 14]
```diff
@@ -127,6 +127,64 @@
 	throw_range = 7
 	force = 0
 	var/random_color = TRUE
+	var/world_state
+	var/storage_state
+	var/current_color
```
Comment (Jacquerel) [inline code/game/objects/items/toys.dm:132]: Also please add documenting comments to these vars (written above the var, prefixed with `///`, describe what the variable is for).

**Review by Jacquerel — CHANGES_REQUESTED** [review]

File: code/game/objects/items/toys.dm (hunk 13 above)
Comment (Jacquerel) [reply code/game/objects/items/toys.dm:163]: Move this list definition from inside this proc to being a property of the balloon object.
And make it `var/static/list/balloon_combos`

File: code/game/objects/items/toys.dm [hunk 15]
```diff
@@ -127,6 +127,64 @@
 	throw_range = 7
 	force = 0
 	var/random_color = TRUE
+	var/world_state //the string of the dmi state the balloon has while on the floor.
+	var/storage_state //the string of the dmi state the balloon has while in your inventory.
+	var/current_color //the string describing the name of balloon's current colour.
```
Comment (Jacquerel) [inline code/game/objects/items/toys.dm:132]:
```suggestion
	/// the string of the dmi state the balloon has while on the floor.
	var/world_state 
	/// the string of the dmi state the balloon has while in your inventory.
	var/storage_state 
	/// the string describing the name of balloon's current colour.
	var/current_color 
```

File: code/game/objects/items/toys.dm [hunk 16]
```diff
@@ -127,6 +127,64 @@
...
+		list("blue", "yellow") = /obj/item/toy/balloon_animal/ai,
+		list("blue", "orange") = /obj/item/toy/balloon_animal/dog,
+		list("blue", "purple") = /obj/item/toy/balloon_animal/xeno,
+		list("green", "yellow") = /obj/item/toy/balloon_animal/banana,
+		list("green", "orange") = /obj/item/toy/balloon_animal/lizard,
+		list("green", "purple") = /obj/item/toy/balloon_animal/slime,
+		list("yellow", "orange") = /obj/item/toy/balloon_animal/moth,
+		list("yellow", "purple") = /obj/item/toy/balloon_animal/ethereal,
+		list("orange", "purple") = /obj/item/toy/balloon_animal/plasmaman,
+	) /// For balloon animals.
```
Comment (Jacquerel) [inline code/game/objects/items/toys.dm:156]:
```suggestion
	/// Combinations of balloon colours to make specific animals.
	var/list/balloon_combos = list(
		list("red", "blue") = /obj/item/toy/balloon_animal/guy,
		list("red", "green") = /obj/item/toy/balloon_animal/nukie,
		list("red", "yellow") = /obj/item/toy/balloon_animal/clown,
		list("red", "orange") = /obj/item/toy/balloon_animal/cat,
		list("red", "purple") = /obj/item/toy/balloon_animal/fly,
		list("blue", "green") = /obj/item/toy/balloon_animal/podguy,
		list("blue", "yellow") = /obj/item/toy/balloon_animal/ai,
		list("blue", "orange") = /obj/item/toy/balloon_animal/dog,
		list("blue", "purple") = /obj/item/toy/balloon_animal/xeno,
		list("green", "yellow") = /obj/item/toy/balloon_animal/banana,
		list("green", "orange") = /obj/item/toy/balloon_animal/lizard,
		list("green", "purple") = /obj/item/toy/balloon_animal/slime,
		list("yellow", "orange") = /obj/item/toy/balloon_animal/moth,
		list("yellow", "purple") = /obj/item/toy/balloon_animal/ethereal,
		list("orange", "purple") = /obj/item/toy/balloon_animal/plasmaman,
	)
```

**Review by Jacquerel — CHANGES_REQUESTED** [review]

**Comment by ShizCalev** [comment]
What in the fuck.

**Pull request merged**